posts/Gunnar/2017-01-31-preventing-leaky-apis-with-jqassistant.adoc

```diff
@@ -0,0 +1,227 @@
+                    </configuration>
+                </execution>
+            </executions>
+        </plugin>
+    </plugins>
+</build>
+...
+----
+
+With the plug-in in place, we can define a Cypher query which finds any API method with an internal return type:
+
+[source,xml]
+._jqassistant/rules.xml_
+----
+<?xml version="1.0" encoding="UTF-8"?>
+<jqa:jqassistant-rules xmlns:jqa="http://www.buschmais.com/jqassistant/core/analysis/rules/schema/v1.0">
+
+    <constraint id="my-rules:PublicMethodsMayNotExposeInternalTypes">
+        <description>API/SPI methods must not expose internal types.</description>
+        <cypher><![CDATA[
+            MATCH
+                (class)-[:`DECLARES`]->(method)-[:`RETURNS`]->(returntype)
+            WHERE
+                NOT class.fqn =~ ".*\\.internal\\..*"
+                AND (method.visibility="public" OR method.visibility="protected")
+                AND returntype.fqn =~ ".*\\.internal\\..*"
+            RETURN
+                method
+        ]]></cypher>
+    </constraint>
+
+    <group id="default">
+        <includeConstraint refId="my-rules:PublicMethodsMayNotExposeInternalTypes" />
+    </group>
+
+</jqa:jqassistant-rules>
+----
+
+jQAssistant rules are given in a file named _jqassistant/rules.xml_ by default.
+
+The rules themselves are defined using the Cypher query language.
+If you haven't used Cypher before, it may feel a bit uncommon at first,
+but from my own experience I can tell you that one gets the hang of it pretty quickly.
+Essentially, you describe patterns of graph nodes, their properties, their type (as expressed via "labels", a kind of tag) and their relationships.
+
+In the query above, we use a pattern including three nodes ("class", "method" and "returntype") and two relationships.
+There must be a relationship of type "DECLARES" from the "class" node to the "method" node
+and another relationship of type "RETURNS" from the "method" to the "returntype" node.
+
+As we are only interested in methods leaking internal types through the API,
+the `WHERE` clause is used to further refine the selection:
+
+* The declaring class must be part of the API (it must not be in an `internal` package, as expressed by filtering on the fully-qualified class name),
+* the method must either have public or protected visibility and
+* the return type must be located in an `internal`.
+
+To execute the rule, simply build the project with the jQAssistant Maven plug-in configured as above.
+The plug-in will automatically execute all rules of the default group given in the rules file.
+If there is any result for any of the executed rules, the build will fail, displaying the result(s) of the affected rules:
+
+[source]
+----
+[INFO] --- jqassistant-maven-plugin:1.1.4:analyze (default) @ hibernate-validator ---
+...
+[ERROR] --[ Constraint Violation ]-----------------------------------------
+[ERROR] Constraint: my-rules:PublicMethodsMayNotExposeInternalTypes
+[ERROR] Severity: INFO
```

Comment:
I was a bit surprised that a rule with INFO severity would fail the build when I first tested it. Wondering if we should change it to something more in line with what it is. It's a detail though.